edgedb-tokio/src/builder.rs

```diff
@@ -464,34 +804,70 @@ impl Builder {
     /// instead if possible.
     ///
     /// This will mark the builder as initialized (if reading is successful)
-    /// and overwrite all credentials. However, `insecure_dev_mode`, pools
+    /// and overwrite all credentials. However, `client_security`, pools
     /// sizes, and timeouts are kept intact.
-    #[cfg(feature="fs")]
     pub async fn read_instance(&mut self, name: &str)
         -> Result<&mut Self, Error>
     {
-        if !is_valid_instance_name(name) {
-            return Err(ClientError::with_message(format!(
-                "instance name {:?} contains unsupported characters", name)));
+        let name = InstanceName::from_str(name)?;
+        match &name {
+            InstanceName::Local(name) => {
+                self.read_credentials(
+                    config_dir()?
+                        .join("credentials")
+                        .join(format!("{}.json", name)),
+                )
+                .await?;
+            }
+            InstanceName::Cloud { org_slug, name } => {
+                let secret_key = if let Some(secret_key) = &self.secret_key {
+                    secret_key.clone()
+                } else {
+                    let profile = self
+                        .cloud_profile
+                        .as_deref()
+                        .map(|s| Ok(s.to_string()))
+                        .or_else(|| get_env("EDGEDB_CLOUD_PROFILE").transpose())
+                        .unwrap_or_else(|| Ok(String::from("default")))?;
+                    let path = cloud_config_file(&profile)?;
+                    let data = fs::read(path).await
+                        .map_err(ClientError::with_source)?;
+                    let config: CloudConfig = from_slice(&data)
+                        .map_err(ClientError::with_source)?;
+                    config.secret_key
+                };
+                let claims_b64 = secret_key
+                    .splitn(3, ".")
+                    .skip(1)
+                    .next()
+                    .ok_or(ClientError::with_message("Illegal JWT token"))?;
+                let claims = base64::decode_config(claims_b64, base64::URL_SAFE_NO_PAD)
+                    .map_err(ClientError::with_source)?;
+                let claims: Claims = from_slice(&claims)
+                    .map_err(ClientError::with_source)?;
+                let dns_zone = claims
+                    .issuer
+                    .unwrap_or_else(|| EDGEDB_CLOUD_DEFAULT_DNS_ZONE.to_string());
```

Comment:
I missed out this change I think:

```suggestion
                    .ok_or(ClientError::with_message("Invalid secret key"))?;
```